app/controllers/devise/two_factor_authentication_controller.rb

```diff
@@ -2,16 +2,50 @@ class Devise::TwoFactorAuthenticationController < DeviseController
   prepend_before_action :authenticate_scope!
   before_action :prepare_and_validate, :handle_two_factor_authentication
 
+  OTP_CODE = 1
+  TOTP_CODE = 2
+  BACKUP_CODE = 3
+
   def show
   end
 
+  def show_backup
+  end
+
+  def show_totp
+  end
+
+  def use_backup_code
+    render :show_backup
+  end
+
+  def use_totp_code
+    render :show_totp
+  end
+
   def update
-    render :show and return if params[:code].nil?
+    type = if !params[:code].nil?
+             OTP_CODE
+           elsif !params[:totp_code].nil?
+             TOTP_CODE
+           elsif !params[:backup_code].nil?
+             BACKUP_CODE
+           else
+             render :show and return
+           end
+
+    auth_status = if type == BACKUP_CODE
+                    resource.authenticate_backup_code(params[:backup_code])
+                  elsif type == OTP_CODE
+                    resource.authenticate_otp(params[:code])
+                  elsif type == TOTP_CODE
+                    resource.authenticate_totp(params[:totp_code])
+                  end
 
-    if resource.authenticate_otp(params[:code])
+    if auth_status
       after_two_factor_success_for(resource)
     else
-      after_two_factor_fail_for(resource)
+      after_two_factor_fail_for(resource, type)
     end
```

Comment:
This method is doing too many things. It violates the Single Responsibility Principle. If you ran Rubocop against this method, it would raise several offenses. One way to find out if a class or method is doing too much is to describe it in plain English. In this case, you might say that this class verifies 3 different methods of two factor authentication. Sometimes it verifies a direct OTP, sometimes it verifies a TOTP, and sometimes it verifies a backup code. That's 3 different things this class is responsible for.

I would recommend moving each method of verification into its own controller.